File: 3rd_party_slots/rust_metta_bus_client/src/bus.rs
use std::collections::HashMap;

pub static PATTERN_MATCHING_QUERY: &str = "pattern_matching_query";
pub static QUERY_EVOLUTION: &str = "query_evolution";

#[derive(Debug, Default, Clone)]
pub struct Bus {
	pub command_owner: HashMap<String, String>,
}

impl Bus {
	pub fn new() -> Self {
		Default::default()
	}

	pub fn add(&mut self, command: String) {
		if let Some(owner) = self.command_owner.get(&command) {
			if !owner.is_empty() {
				panic!("Bus: command <{command}> is already assigned to {owner}");
			}
		} else {
			self.command_owner.insert(command, String::new());
		}
	}

	pub fn set_ownership(&mut self, command: String, node_id: &str) {
		if !self.command_owner.contains_key(&command) {
			panic!("Bus: command <{command}> is not defined");
		} else if self.command_owner[&command].is_empty() {
			self.command_owner.insert(command.to_string(), node_id.to_string());
		} else if self.command_owner[&command] != node_id {
			panic!(
				"Bus: command <{command}> is already assigned to {}",
				self.command_owner[&command]
			);
		}
	}

	pub fn get_ownership(&self, command: String) -> &str {
		match self.command_owner.get(&command) {
			Some(owner) => owner,
			None => panic!("Bus: unknown command <{command}>"),
		}
	}

	pub fn contains(&self, command: String) -> bool {
		self.command_owner.contains_key(&command)
	}
}

File: 3rd_party_slots/python_client/hyperon_das/bus.py
import enum


class BusCommand(str, enum.Enum):
    PATTERN_MATCHING_QUERY = "pattern_matching_query"


class Bus:
    def __init__(self, other: 'Bus' = None) -> None:
        if other is not None:
            self._command_owner = dict(other._command_owner)
        else:
            self._command_owner = {}

    def add(self, command: str) -> None:
        if command in self._command_owner:
            if self._command_owner[command] != "":
                raise RuntimeError(f"Bus: command <{command}> is already assigned to {self._command_owner[command]}")
        else:
            self._command_owner[command] = ""

    def set_ownership(self, command: str, node_id: str) -> None:
        if command not in self._command_owner:
            raise RuntimeError(f"Bus: command <{command}> is not defined")
        else:
            if self._command_owner[command] == "":
                self._command_owner[command] = node_id
            else:
                if self._command_owner[command] != node_id:
                    raise RuntimeError(f"Bus: command <{command}> is already assigned to {self._command_owner[command]}")

    def get_ownership(self, command: str) -> str:
        if command not in self._command_owner:
            raise RuntimeError(f"Bus: unknown command <{command}>")
        return self._command_owner[command]

    def contains(self, command: str) -> bool:
        return command in self._command_owner

    def to_string(self) -> str:
        entries = []
        for cmd, owner in self._command_owner.items():
            if owner:
                entries.append(f"{cmd}:{owner}")
            else:
                entries.append(cmd)
        return "{" + ", ".join(entries) + "}"

    def __eq__(self, other: 'Bus') -> bool:
        return self._command_owner == other._command_owner

    def __add__(self, command: str) -> 'Bus':
        self.add(command)
        return self
